### ISSUE TITLE
can't find negative loss in CDN

### ISSUE THREAD

**Issue by Vallum** [IDEA-Research/DINO#119]
Dear authors,
Thank you for opening the fantastic project!

BTW, I am wondering where I can find the negative loss of CDN.
I could see that you used the only positive cases for losses at https://github.com/IDEA-Research/DINO/blob/f7262ddcbc5a9fcf4cd2ac9924cee37cdb459a05/models/dino/dino.py#L518
```
        if self.training and dn_meta and 'output_known_lbs_bboxes' in dn_meta:
            output_known_lbs_bboxes,single_pad, scalar = self.prep_for_dn(dn_meta)

            dn_pos_idx = []
            dn_neg_idx = []
            for i in range(len(targets)):
                if len(targets[i]['labels']) > 0:
                    t = torch.range(0, len(targets[i]['labels']) - 1).long().cuda()
                    t = t.unsqueeze(0).repeat(scalar, 1)
                    tgt_idx = t.flatten()
                    output_idx = (torch.tensor(range(scalar)) * single_pad).long().cuda().unsqueeze(1) + t
                    output_idx = output_idx.flatten()
                else:
                    output_idx = tgt_idx = torch.tensor([]).long().cuda()

                dn_pos_idx.append((output_idx, tgt_idx))
                dn_neg_idx.append((output_idx + single_pad // 2, tgt_idx))


            output_known_lbs_bboxes=dn_meta['output_known_lbs_bboxes']
            l_dict = {}
            for loss in self.losses:
                kwargs = {}
                if 'labels' in loss:
                    kwargs = {'log': False}
                l_dict.update(self.get_loss(loss, output_known_lbs_bboxes, targets, dn_pos_idx, num_boxes*scalar,**kwargs))
```

But it seems like you used only 'dn_pos_idx'.
In that case,  DN-DETR and DINO look like same to me without MQS and LFT.

What did I miss?

**Comment by SuperHenry2333** [comment]
The label loss contains loss for negative samples. It will classify any predictions that are not matched with gt as "no object".

**Comment by sipie800** [comment]
> The label loss contains loss for negative samples. It will classify any predictions that are not matched with gt as "no object".

But your code is weird anyway.

        known_indice = known_indice.repeat(2 * dn_number, 1).view(-1)
        known_labels = labels.repeat(2 * dn_number, 1).view(-1)
        known_bid = batch_idx.repeat(2 * dn_number, 1).view(-1)
        known_bboxs = boxes.repeat(2 * dn_number, 1)
        known_labels_expaned = known_labels.clone()
        known_bbox_expand = known_bboxs.clone()

        if label_noise_ratio > 0:
            p = torch.rand_like(known_labels_expaned.float())
            chosen_indice = torch.nonzero(p < (label_noise_ratio * 0.5)).view(-1)  # half of bbox prob
            new_label = torch.randint_like(chosen_indice, 0, num_classes)  # randomly put a new one here
            known_labels_expaned.scatter_(0, chosen_indice, new_label)
        single_pad = int(max(known_num))

You generate 2*dn_number dn, but with

output_idx = (torch.tensor(range(scalar)) * single_pad).long().cuda().unsqueeze(1) + t

you are taking only the first half of them! 
There are 0.5*0.5=0.25 negative samples in the first half samples, the ratio might be right while the logic is wrong. That's not what you intend to do.

**Comment by sipie800** [comment]
Or I might mistake your logic.

    if training:
        targets, dn_number, label_noise_ratio, box_noise_scale = dn_args
        # positive and negative dn queries
        dn_number = dn_number * 2

That's why we got a magic number *2 here?
